src/python/pants/core/goals/publish.py

```diff
@@ -225,47 +287,77 @@ async def run_publish(
-            prep = "to"
-        else:
-            sigil = console.sigil_failed()
-            status = "failed"
-            prep = "for"
-            exit_code = res.exit_code
+    flattened_processes = list(chain.from_iterable(processes))
+    background_publishes: list[PublishPackages] = [
+        pub for pub in flattened_processes if isinstance(pub.process, Process)
+    ]
+    foreground_publishes: list[PublishPackages] = [
+        pub for pub in flattened_processes if isinstance(pub.process, InteractiveProcess)
+    ]
+    skipped_publishes: list[PublishPackages] = [
+        pub for pub in flattened_processes if pub.process is None
+    ]
+    background_requests: list[Get[FallibleProcessResult]] = []
+    for pub in background_publishes:
+        process = cast(Process, pub.process)
+        # Because this is a publish process, we want to ensure we don't cache this process.
+        assert process.cache_scope == ProcessCacheScope.PER_SESSION
+        background_requests.append(
+            Get(
+                FallibleProcessResult,
+                {process: Process, local_environment.val: EnvironmentName},
+            )
+        )
 
+    # Process all non-interactive publishes
+    logger.debug(f"Awaiting {len(background_requests)} background publishes")
+    background_results = await MultiGet(background_requests)
+    for pub, background_res in zip(background_publishes, background_results):
+        logger.debug(f"Processing {pub.process} background process")
+        pub_results, pub_output = _to_publish_output_results_and_data(pub, background_res, console)
+        results.extend(pub_results)
+        outputs.extend(pub_output)
+
+        names = "'" + "', '".join(pub.names) + "'"
+        output_msg = f"Output for publishing {names}"
+        if background_res.stdout:
+            output_msg += f"\n{background_res.stdout.decode()}"
+        if background_res.stderr:
+            output_msg += f"\n{background_res.stderr.decode()}"
+
+        if publish.noninteractive_process_output == ShowOutput.ALL or (
+            publish.noninteractive_process_output == ShowOutput.FAILED
+            and background_res.exit_code == 0
+        ):
+            console.print_stdout(output_msg)
+
+        if background_res.exit_code != 0:
+            exit_code = background_res.exit_code
+
+    for pub in skipped_publishes:
+        sigil = console.sigil_skipped()
+        status = "skipped"
         if pub.description:
-            status += f" {prep} {pub.description}"
-
+            status += f" {pub.description}"
         for name in pub.names:
             results.append(f"{sigil} {name} {status}.")
+        outputs.append(pub.get_output_data(published=False, status=status))
 
-        outputs.append(
-            pub.get_output_data(
-                exit_code=res.exit_code,
-                published=res.exit_code == 0,
-                status=status,
-            )
+    # Process all interactive publishes
```

Comment:
Might there be code sharing opportunities with deploy.py? Probably for another PR though.